docs: expand README.md, revamp CONTRIBUTING.md, add AGENTS.md (#489)
canonical/charmlibs: AGENTS.md

When writing or editing docstrings in `__init__.py` or other public modules, remember they appear verbatim in the published reference at [documentation.ubuntu.com/charmlibs](https://documentation.ubuntu.com/charmlibs). Keep them informative for library users, not implementation notes.

## Common pitfalls

- **Don't call `uv add` directly** — use `just add <package> <dep>` to respect repo-level constraints.
- **Don't bump the package you're working on to a non-dev version without updating `CHANGELOG.md`** — CI will block the merge.
- **Making changes to multiple libraries in the same PR** — the CI matrix runs per changed package, and review requirements are determined by `CODEOWNERS`.
- **Don't add unnecessary features or refactor code beyond what's asked** — this is a multi-team monorepo with careful versioning; unintended public API changes require major version bumps.

## External resources

| Resource | URL |
|----------|-----|
| charmlibs docs | .docs/index.md |
| ops (charm framework) | https://documentation.ubuntu.com/ops/ |
| ops.testing reference | https://documentation.ubuntu.com/ops/latest/reference/ops-testing/ |

> | Juju docs | https://documentation.ubuntu.com/juju/3.6/ |

Author: ```suggestion
| Juju docs | https://documentation.ubuntu.com/juju/latest/llms.txt |
```

| Charmcraft docs | https://documentation.ubuntu.com/charmcraft/stable/ |
| Jubilant (integration test client) | https://documentation.ubuntu.com/jubilant/ |
| Pebble | https://documentation.ubuntu.com/pebble/ |

> | Concierge (CI Juju setup) | https://github.com/canonical/concierge |

Author: ```suggestion
| Concierge (CI Juju setup) | https://raw.githubusercontent.com/canonical/concierge/refs/heads/main/README.md |
```

# Agent instructions for charmlibs

## Overview

> `charmlibs` is a monorepo of Python libraries for [Juju](https://documentation.ubuntu.com/juju/3.6/) charms.

Reviewer: I feel like this should be latest, rather than 3.6. Unless we want to like to something higher level, like the main Juju page, or Juju docs page. But for an agent, one of the ones that has llms.txt is probably best. You could even link directly there to avoid the trip to find out about the markdown.

```suggestion
`charmlibs` is a monorepo of Python libraries for [Juju](https://documentation.ubuntu.com/juju/latest/) charms.
```

Author: Ah I missed this 3.6 link in the `AGENTS.md`, thanks. I reckon dropping this link entirely is the way to go, and I'll make sure the link in the table at the end isn't 3.6 specific.

```suggestion
`charmlibs` is a monorepo of Python libraries for Juju charms.
```

**Key concepts:**

- **Juju** — Canonical's open-source orchestration engine. Deploys and manages applications on Kubernetes and bare-metal/VM clouds.

> - **Charm** — a Python program that uses the [ops](https://documentation.ubuntu.com/ops/) framework to respond to Juju lifecycle events and manage a workload. Charms can be Kubernetes-based (using [Pebble](https://documentation.ubuntu.com/pebble/) to manage container processes) or machine-based.

Reviewer: Similarly, maybe these could link to the markdown pages once the Sphinx Stack PR is merged?

- **Juju relation** — a named connection between two charms, backed by shared key-value stores called *databags*. Relations are how charms communicate and exchange structured configuration.
- **Charm library** — a reusable Python package that charm authors import. All libraries in this repo are distributed as Python packages on PyPI (not Charmhub-hosted single-file modules, which are a legacy distribution method).

**Two library categories:**

| Category | Namespace | Example package | Import |
|----------|-----------|-----------------|--------|
| General | `charmlibs` | `charmlibs-pathops` | `from charmlibs import pathops` |
| Interface | `charmlibs.interfaces` | `charmlibs-interfaces-tls-certificates` | `from charmlibs.interfaces import tls_certificates` |

> General libraries live at the repo root (e.g. `pathops/`). Interface libraries live under `interfaces/` (e.g. `interfaces/tls-certificates/`).

Reviewer: I know it's for a machine, but I think we should stick with our docs style:

```suggestion
General libraries live at the repo root (e.g. `pathops/`). Interface libraries live under `interfaces/` (for example, `interfaces/tls-certificates/`).
```

Author: Sure -- surely they both tokenise pretty efficiently anyway :)

> ## Repo structure

Reviewer: I'm not really convinced these add value in an AGENTS.md file. I assume an agent made this for you. They also tend to get outdated and the agent can just walk the folder to find things.

Maybe all the comments add the value? If you think so, I'm fine leaving it here.

Author: An agent did make this. I think the comments do add the value though, otherwise the agent needs to dip into the files (for the top-level ones), or scan multiple projects to discover test layout, etc.

├── .template/              # Cookiecutter template used by `just init`
├── .github/workflows/      # CI: ci.yaml, test-package.yaml, test-interface.yaml, publish.yaml
├── justfile                # Task runner recipes
├── docs.just               # Docs-specific recipes
├── interface.just          # Interface-specific recipes
├── pyproject.toml          # Repo-level config: ruff, coverage, shared linting settings
└── test-requirements.txt   # Pinned versions of pytest, coverage, and other test tools
```

## Development workflow

### Prerequisites

`uv` and `just` are required. Verify they're available before proceeding:

```bash
uv --version && just --version
```

If either is missing, they can be installed like this:

```bash
sudo snap install --classic astral-uv   # installs uv
uv tool install rust-just               # installs just (once uv is available)
```

Run `just` or `just help` from anywhere in the repo to see all available commands. All `just` commands execute from the repo root regardless of where they're invoked.

### Inner loop

The command you'll run most often is:

```bash
just check <package>
```

> This runs `just lint <package>`, `just unit <package>`, and `just docs html <package>`. **Run this before every commit on the affected package.**

Reviewer: What about forcing this pre-commit instead? That also helps humans, and it's more reliable because as long as the agent ensures pre-commit is installed (for new possibly ephemeral environments -- for local dev ones typically it already will be) it always happens, rather than hoping this stays prominent in the context.

Or is that not possible because of the mono-repo? If you leave out package is it slow or not checking everything?

Author: Yeah I don't think pre-commit lets you customise which checks to run on which commits does it? So this would mean running every libraries unit tests (and I guess building every library's reference docs (noticeably slow)), etc.

The `<package>` argument is the path from the repo root, e.g. `pathops` or `interfaces/tls-certificates`.

> ### Command reference

Reviewer: Can this come from a `just help` or similar? Generally, it's better to have one or two critical commands and then tell the agent how to get help for everything else.

Author: That's a good idea. I'll keep the table here for now as I don't think the current help is quite up to scratch, but I've added [a comment](https://github.com/canonical/charmlibs/issues/405#issuecomment-4531720129) to the "refactor the justfile" rainy day item about dropping this table.

| `just integration-machine <package>` | Run Juju integration tests against machine, excluding `pytest.mark.k8s_only` tests |
| `just docs html [packages]` | Build docs (including reference docs for all packages by default, or named ones -- run `just docs html -` to exclude all package reference docs (faster when working on docs only) |
| `just docs` | Alias for `just docs html` |
| `just add <package> <dep>` | Add a dependency to a library |
| `just init` | Scaffold a new general library |
| `just interface init` | Scaffold a new interface library |

Extra arguments to `just unit`, `just functional`, and `just integration-*` are passed through to pytest:

```bash
just unit pathops -x -k test_copy   # stop on first failure, filter by name
```

### Adding dependencies

**Always use `just add <package> <dep>` instead of calling `uv add` directly.** This applies repo-level version constraints from `test-requirements.txt`, which is necessary to keep the lockfile consistent:

```bash
just add pathops 'pydantic>=2'
just add interfaces/tls-certificates --requirements my-requirements.txt
```

> ### IDE setup

Reviewer: This sounds like a human instruction? What IDE does an agent have?

Author: True, removed.

Point your IDE to `<package>/.venv` after running any `just` test command. The virtual environment is created there automatically.

> 

Author: ```suggestion
```

## Test types

| Type | Command | What it tests | External requirements |
|------|---------|---------------|----------------------|
| Unit | `just unit <package>` | Logic with mocked externals | None |

> | Functional | `just functional <package>` | Interaction with real external processes (not Juju) | Varies (e.g. pebble, sudo) |

Reviewer: ```suggestion
| Functional | `just functional <package>` | Interaction with real external processes (not Juju) | Varies (for example, pebble, sudo) |
```

| Integration | `just integration-k8s/machine <package>` | Library in a real Juju deployment | charmcraft + Juju controller |

A test type is only executed if the corresponding `tests/` subdirectory exists. Remove a directory to skip that test type entirely.

> Read more: [types of tests in the charmlibs monorepo](https://documentation.ubuntu.com/charmlibs/explanation/charmlibs-tests/).

Reviewer: Again could point to markdown when this repo has Sphinx Stack.

## Commit and PR conventions

> **PRs are squash-merged.** The PR title becomes the single commit message on `main`. Branch commit messages don't need to follow any format.

Reviewer: I think we let humans go wild with their commit style, but we might as well give agents more of a structure. Otherwise, they tend to read through the Git log to match what's there anyway, and it's nice to avoid that.

Author: ```suggestion
**PRs are squash-merged.** The PR title becomes the single commit message on `main`. Branch commit messages are for local reference, and should follow conventional commits for clarity.
```

> PR titles must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). When a PR affects a single library, use the distribution package name without the leading `charmlibs-` as the scope:

Reviewer: From the sound of this, we're blessing having agents open PRs, not just commit. Are you ok with that?

Author: I'm somewhat open -- it's not how I do it, but I don't want to block library owners from running things that way.

```
feat(pathops): add copytree and rmtree
fix(apt): handle multiarch package names correctly
chore(interfaces-tls-certificates): update to Pydantic v2
docs: improve tutorial for adding integration tests
```

**One PR should normally touch only one library.** The CI uses changed files to determine which packages to test and, on merge, which packages to publish.

### Versioning

- Libraries use semantic versioning (`MAJOR.MINOR.PATCH`).
- Dev versions (`X.Y.Z.devN`) are excluded from release CI — safe for in-progress work.
- When bumping to a non-dev version, you **must** also update `CHANGELOG.md`. CI will block the merge otherwise.
- The CI automatically publishes to PyPI on merge when a non-dev version bump is detected.

## Package anatomy

A typical general library looks like this:

```
pathops/
├── src/charmlibs/pathops/
│   ├── __init__.py         # Exports public API; module docstring appears in reference docs
│   ├── _pathops.py         # Private implementation module(s)
│   └── _version.py         # __version__ string (auto-generated)
├── tests/
│   ├── unit/
│   │   ├── conftest.py
│   │   └── test_*.py
│   ├── functional/

> │   │   ├── setup.sh        # Sourced before tests (e.g. start pebble)

Reviewer: ```suggestion
│   │   ├── setup.sh        # Sourced before tests (for example, start pebble)
```

> │   │   ├── teardown.sh     # Sourced after tests (e.g. kill pebble)

Reviewer: ```suggestion
│   │   ├── teardown.sh     # Sourced after tests (for example, kill pebble)
```

├── uv.lock                 # Locked dependencies — commit this
├── README.md
└── CHANGELOG.md            # Must be updated before each release
```

The source uses a namespace package structure: `src/charmlibs/<name>/`. Public API is exported from `__init__.py` via `__all__`. Implementation lives in private submodules (prefixed with `_`). The module docstring in `__init__.py` appears as the top-level description in the generated reference docs.

## Interface libraries

Interface libraries manage the structured data that charms exchange over a Juju relation databag. Key differences from general libraries:

- Live under `interfaces/<interface-name>/`, named exactly as the interface name appears in `charmcraft.yaml`.
- Source under `src/charmlibs/interfaces/<interface_name>/`.
- Usually include a `testing/` subdirectory with a separate `charmlibs-interfaces-<name>-testing` package. This provides `relation_for_provider()` and `relation_for_requirer()` helpers for charm unit tests.
- Typically have unit and integration tests but no functional tests (all meaningful interaction is through Juju).
- Use `just interface init` to scaffold.

> Read more: [how to design relation interfaces](https://documentation.ubuntu.com/charmlibs/how-to/design-relation-interfaces/), [how to provide relation data for charm tests](https://documentation.ubuntu.com/charmlibs/how-to/provide-relation-data-for-charm-tests/).

Reviewer: More stuff to point at markdown versions.

Reference docs are auto-generated from Python docstrings via Sphinx autodoc. The docs source lives in `.docs/`. Build reference docs for a specific package with:

```bash
just docs html pathops
```

This is also run as part of `just check`. The full docs site (all packages) is built with `just docs`.

When writing or editing docstrings in `__init__.py` or other public modules, remember they appear verbatim in the published reference at [documentation.ubuntu.com/charmlibs](https://documentation.ubuntu.com/charmlibs). Keep them informative for library users, not implementation notes.

## Common pitfalls

- **Don't call `uv add` directly** — use `just add <package> <dep>` to respect repo-level constraints.
- **Don't bump the package you're working on to a non-dev version without updating `CHANGELOG.md`** — CI will block the merge.
- **Making changes to multiple libraries in the same PR** — the CI matrix runs per changed package, and review requirements are determined by `CODEOWNERS`.
- **Don't add unnecessary features or refactor code beyond what's asked** — this is a multi-team monorepo with careful versioning; unintended public API changes require major version bumps.

> ## External resources

Reviewer: I think most of these could be to markdown versions, the Juju one could be to latest, and *maybe* Concierge should be to rawgithubcontent of the readme?

Author: Good call, I'll have to see which of these have markdown available. For charmlibs itself, probably referring to the local files makes more sense.

Author: Looks like only Juju for now. Good call on the raw markdown for Concierge.

| Resource | URL |
|----------|-----|

> | charmlibs docs | https://documentation.ubuntu.com/charmlibs |

Author: ```suggestion
| charmlibs docs | .docs/index.md |
```
